video/out/vo_vaapi_wayland.c

```diff
@@ -0,0 +1,371 @@
+	struct priv *p = vo->priv;
+	struct vo_wayland_state *wl = vo->wl;
+	VAStatus status;
+	struct va_pool_entry *entry = calloc(1, sizeof(struct va_pool_entry));
+	if (!entry)
+		return NULL;
+
+	/* extract dmabuf surface descriptor */
+	entry->surface = va_surface_id(src);
+	status = vaExportSurfaceHandle(p->display, entry->surface,
+			VA_SURFACE_ATTRIB_MEM_TYPE_DRM_PRIME_2,
+			VA_EXPORT_SURFACE_COMPOSED_LAYERS | VA_EXPORT_SURFACE_READ_ONLY,
+			&entry->desc);
+	if (status == VA_STATUS_ERROR_INVALID_SURFACE) {
+		MP_VERBOSE(vo, "VA export to composed layers not supported.\n");
+		va_free_entry(entry);
+		return NULL;
+	} else if (!vo_wayland_supported_format(vo,
+			entry->desc.layers[0].drm_format)) {
+		MP_VERBOSE(vo, "%s not supported.\n",
+				vo_wayland_drm_format_string(entry->desc.layers[0].drm_format));
+		va_free_entry(entry);
+		return NULL;
+	} else if (!CHECK_VA_STATUS(vo, "vaExportSurfaceHandle()")) {
+		va_free_entry(entry);
+		return NULL;
+	} else {
+		uint i, j, plane = 0;
+
+		entry->params = zwp_linux_dmabuf_v1_create_params(wl->dmabuf);
+		for (i = 0; i < entry->desc.num_layers; i++) {
+			entry->drm_format = entry->desc.layers[i].drm_format;
+			for (j = 0; j < entry->desc.layers[i].num_planes; ++j) {
+				int object = entry->desc.layers[i].object_index[j];
+				uint64_t modifier =
+						entry->desc.objects[object].drm_format_modifier;
+				zwp_linux_buffer_params_v1_add(entry->params,
+						entry->desc.objects[object].fd, plane,
+						entry->desc.layers[i].offset[j],
+						entry->desc.layers[i].pitch[j], modifier >> 32,
+						modifier & 0xffffffff);
+				plane++;
+			}
+		}
+	}
+	entry->buffer = zwp_linux_buffer_params_v1_create_immed(entry->params,
+			src->params.w, src->params.h, entry->drm_format, 0);
+
+	return entry;
+}
+static void va_pool_free(struct va_pool *pool) {
+	if (!pool)
+		return;
+
+	for (uint i = 0; i < pool->num_entries; ++i)
+		va_free_entry(pool->entries[i]);
+	free(pool->entries);
+	free(pool);
+}
+static struct va_pool* va_pool_alloc(struct vo *vo) {
+	struct va_pool *pool = (struct va_pool*) calloc(1, sizeof(struct va_pool));
+
+	if (!pool)
+		return NULL;
+	pool->num_allocated = 30;
+	pool->entries = (struct va_pool_entry**) calloc(pool->num_allocated,
+			sizeof(struct va_pool_entry*));
+	if (!pool->entries) {
+		va_pool_free(pool);
+		return NULL;
+	}
+	pool->vo = vo;
+
+	return pool;
+}
+static struct va_pool_entry* va_pool_alloc_entry(struct va_pool *pool,
+		struct mp_image *src) {
+	VASurfaceID surface;
+
+	if (!pool)
+		return NULL;
+
+	surface = va_surface_id(src);
+	for (uint i = 0; i < pool->num_entries; ++i) {
+		struct va_pool_entry *item = pool->entries[i];
+		if (item->surface == surface)
+			return pool->entries[i];
+	}
+
+	struct va_pool_entry *entry = va_alloc_entry(pool->vo, src);
+	if (!entry)
+		return NULL;
+	if (pool->num_entries == pool->num_allocated) {
+		uint current_num_allocated = pool->num_allocated;
+		struct va_pool_entry **items = realloc(pool->entries,
```

Comment:
Not too sure about some of these `realloc/calloc` usages that creep up in here. Could you rework it so it uses talloc in some way? I know these entry objects have stuff in there that make it not quite as simple as a normal free, so it might not work out exactly.